contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,164 @@
+  // The time at which these Metrics were updated.  
+  Timestamp metav1.Time  
+  // Cumulative CPU usage (sum of all cores) since object creation.  
+  CumulativeUsageNanoSeconds *uint64   
+}  
+
+// MemoryUsage holds statistics about the quantity of memory consumed  
+type MemoryUsage struct {  
+  // The time at which these metrics were updated.  
+  Timestamp metav1.Time  
+  // The amount of "working set" memory. This includes recently accessed memory,  
+  // dirty memory, and kernel memory.  
+  UsageBytes *uint64  
+}   
+
+// FilesystemUsage holds statistics about the quantity of local storage (e.g. disk) resources consumed  
+type FilesystemUsage struct {  
+  // The time at which these metrics were updated.  
+  Timestamp metav1.Time  
+  // StorageIdentifier must uniquely identify the node-level storage resource that is consumed.
+  // It may utilize device, partition, filesystem id, or other identifiers.  
+  StorageIdentifier string  
+  // UsedBytes represents the disk space consumed, in bytes.  
+  UsedBytes *uint64  
+  // UsedInodes represents the inodes consumed  
+  UsedInodes *uint64  
+}  
+```
+
+### On-Demand Design
+Interface:  
+The interface for exposing these metrics within the kubelet contains methods for fetching each relevant metric.  These methods contains a "recency" parameter which specifies how recently the metrics must have been computed.  Kubelet components which require very up-to-date metrics (eviction, for example), use very low values.  Other components use higher values.  
+
+Implementation:  
+To keep performance bounded while still offering metrics "On-Demand", all calls to get metrics are cached, and a minimum recency is established to prevent repeated metrics computation.  Before computing new metrics, the previous metrics are checked to see if they meet the recency requirements of the caller.  If the age of the metrics meet the recency requirements, then the cached metrics are returned.  If not, then new metrics are computed and cached.  
+
+## Implementation Plan 
+@dashpole will modify the structure of metrics collection code to be "On-Demand".   
+
+Suggested, tentative future work, which may be covered by future proposals:  
+ - Publish these metrics in some form to a kubelet API endpoint
+ - Obtain all runtime-specific information needed to collect metrics from the CRI.   
+ - Kubernetes can be configured to run a default "third party metrics provider" as a daemonset.  Possibly standalone cAdvisor.
+
+## Rollout Plan
+Once this set of metrics is accepted, @dashpole will begin discussions on the format, and design of the endpoint that exposes them.  The node resource metrics endpoint (TBD) will be added alongside the current Summary API in an upcoming release.  This should allow concurrent developments of other portions of the system metrics pipeline (metrics-server, for example).  Once this addition is made, all other changes will be internal, and will not require any API changes.  
```

Comment:
As we discussed offline nobody works on metrics-server in Q1/1.6. Once metrics-server will be the main consumer of the API I think we should wait for it rather than add this in the upcoming release.